topydo/ui/TodoListWidget.py

```diff
@@ -101,132 +120,208 @@ def _scroll_to_bottom(self, p_size):
         self.listbox.calculate_visible(p_size)
 
     def keypress(self, p_size, p_key):
-        # first check whether 'g' was pressed previously
-        if self.keystate == 'g':
-            if p_key == 'g':
-                self._scroll_to_top(p_size)
+        urwid.emit_signal(self, 'remove_pending_action')
+        requires_further_input = ['postpone', 'postpone_s', 'pri']
 
-            # make sure to accept normal shortcuts again
-            self.keystate = None
-            return
-        elif self.keystate in ['p', 'ps']:
-            if p_key not in ['d', 'w', 'm', 'y']:
-                if p_key.isdigit():
-                    self._pp_offset += p_key
-                elif self.keystate == 'p' and p_key == 's':
-                    self.keystate = 'ps'
-                else:
-                    self._pp_offset = ''
-                    self.keystate = None
+        keymap, keystates = self.keymap
+
+        shortcut = self.keystate or ''
+        shortcut += translate_key_to_config(p_key)
+
+        try:
+            action = keymap[shortcut]
+        except KeyError:
+            action = None
+
+        if action:
+            if shortcut in keystates:
+                # Supplied key-shortcut matches keystate and action. Save the
+                # keystate in case user will hit another key and add an action
+                # waiting for execution if user won't type anything further.
+                self.keystate = shortcut
```

Comment:
Consider making self.keystate a `@property`, which setter emits `show_keystate`. That would simplify this code a bit.